distributed/shuffle/_pickle.py

```diff
@@ -40,3 +45,98 @@ def unpickle_bytestream(b: bytes | bytearray | memoryview) -> Iterator[Any]:
         if remainder.nbytes == 0:
             break
         b = remainder
+
+
+def pickle_dataframe_shard(
+    input_part_id: int,
+    shard: pd.DataFrame,
+) -> list[pickle.PickleBuffer]:
+    """Optimized pickler for pandas Dataframes. DIscard all unnecessary metadata
+    (like the columns header).
+
+    Parameters:
+        obj: pandas
+    """
+    return pickle_bytelist(
+        (input_part_id, shard.index, *shard._mgr.blocks), prelude=False
+    )
```

Comment:
Fun fact: This PR works fine with cudf-backed data when I do something like the following:

```python
    shard = shard.to_pandas()  if hasattr(shard, "to_pandas") else shard
    return pickle_bytelist(
        (input_part_id, shard.index, *shard._mgr.blocks), prelude=False
    )
```

as long as I also modify the last line of `unpickle_and_concat_dataframe_shards` to be:

```python
    result = dd.methods.concat(shards, copy=True)
    if hasattr(type(meta), "from_pandas"):
        return type(meta).from_pandas(result)
    return result
```

I'm sure there are cases where the cudf->pandas->cudf round trip isn't perfect, but the small amount of logic I needed to work-around the changes is promising.